### DISCUSSION TITLE
Visual Debugger / Custom Benchmark in Godot Editor [possibly a plugin]

### DISCUSSION THREAD

**Discussion by GersonFeDutra** [godotengine/godot-proposals#11421] [Editor]
Something like
[![https://youtu.be/4x9GvQKiV0Y](https://europe1.discourse-cdn.com/unity/original/4X/d/8/1/d81968fbeb686c5df31fa6eced2c2463c3840773.png)](https://youtu.be/4x9GvQKiV0Y)

Could be core-implemented in Gdscript with a `print_graph(min, max, current [, history_size])` method that plost custom [`Variant`] values in a graph placed in a bottom docker - like the profiler; or directly in the output terminal (that already supports visuals, thanks to the `RichTextLabel`).

The desired implementation would be for `print_graph()` be either time-sensitive or time-independent, and update the graph each time `print_graph` was called in the running game. The plot could be an interpolation of the `current`values, where `min` and `max` set the y-values' plot area ("the graph's vertical zoom"), while `history_size` is an optional value that sets the "horizontal zoom".
_A further idea, would be to allow for custom implementations for Resources and custom Resources, with custom interpolation and plot functions._

Inspiration: [C++ Visual Benchmark](https://www.youtube.com/watch?v=xlAH4dbMVnU)

**Comment by GersonFeDutra** [comment]
I had this issue while debugging physics for my game, where printing numbers is both not ideal and not intuitive.

![image](https://github.com/user-attachments/assets/9ac49e0a-515e-4e9b-9f22-a90d28b13ae0)

**Comment by sockeye-d** [comment]
Maybe to make it easier to set up you could have something like this:
```gdscript
create_graph(graph_id: PackedStringArray, time_length: float = 3, min_y: Variant = null, max_y = null)

print_graph(graph_id: String, values: float)

print_graphs(graph_id: String, values: Dictionary[String, float])
```
You would call create_graph to set up the graph area. Multiple graphs can be drawn on the same area by passing in multiple values into `print_graphs` like this:
```gdscript
print_graphs("my_graph", { "value0": value_0, "value1": value_1 })
```

This is also technically possible with an EditorDebuggerPlugin but it takes a lot of setup

**Comment by Calinou** [comment]
You can already create custom graphs that appear in the **Debugger > Monitors** tab of the editor when the project is running using [custom performance monitors](https://docs.godotengine.org/en/stable/tutorials/scripting/debug/custom_performance_monitors.html). While this solution is meant to be used to track performance-related variables, you can technically use it for anything.

The downside is that graphs are only sampled once per second, so they'll lack precision for values that update quickly.

**Reply by GersonFeDutra** [reply]
Tried this solution, but the precision issue got me off.
I am working with a sensible rope system, sometimes I track values just for a couple of milliseconds. It is harder to see the behaviour of vectors.
